File: backend/src/MMseqs/Lookup.php
<?php
namespace MMseqs;

class Lookup {
    static public function read($file, $limit, $page) {
        $handle = fopen($file, "r");
        if ($handle) {
            $i = 0;
            $start = ((int)$page) * $limit;
            $result = [];
            $hasNext = false;
            while (($line = fgets($handle)) !== false) {
                $current = $i++;
                if ($current < $start) {
                    continue;
                }
                if ($current >= ($start+$limit)) {
                    $hasNext = (fgets($handle) !== false);
                    break;
                }
                $values = explode("\t", $line);
                $result[] = [ "id" => ((int) $values[0] - 1), "name" => trim($values[1]) ]; 
            }
            fclose($handle);
            return [ "items" => $result, "hasNext" => $hasNext ];
        } else {
            throw new Exception("Could not read input.lookup!");
        } 
    }
}

File: controller/lookup.go
package controller

import (
	"github.com/go-redis/redis"
	"github.com/satori/go.uuid"

	"os"
	"path/filepath"

	"github.com/milot-mirdita/mmseqs-web-backend/tsv"
)

type LookupResult struct {
	Id   uint32 `json:"id"`
	Name string `json:"name"`
}

type LookupResponse struct {
	Lookup []LookupResult `json:"lookup"`
	HasNextPage bool `json:"hasNext"`
}

func Lookup(client *redis.Client, ticket uuid.UUID, page uint64, limit uint64, basepath string) (LookupResponse, error) {
	res, err := client.Get("mmseqs:status:" + ticket.String()).Result()
	if err != nil {
		return LookupResponse{}, err
	}

	if res == "COMPLETED" {
		result := filepath.Join(basepath, ticket.String(), "input.lookup")

		file, err := os.Open(result)
		defer file.Close()
		if err != nil {
			return LookupResponse{}, err
		}

		var results []LookupResult
		res := LookupResult{}
		parser := tsv.NewParser(file, &res)

		var cnt uint64
		cnt = 0
		start := page * limit
		hasNextPage := false
		for {
			eof, err := parser.Next()
			if eof {
				break
			}
			if err != nil {
				return LookupResponse{}, err
			}

			cnt++
			if cnt < start {
				continue
			}

			if cnt >= (start + limit) {
				if	eof, _ := parser.Next(); !eof {
					hasNextPage = true
				}
				break
			}

			res.Id--
			results = append(results, res)
		}

		return LookupResponse{results, hasNextPage}, nil
	}
	return LookupResponse{}, nil
}
